iree_tests/download_remote_files.py

```diff
@@ -4,45 +4,112 @@
+        if size < 1024.0 or unit == "PiB":
+            break
+        size /= 1024.0
+    return f"{size:.{decimal_places}f} {unit}"
+
+
+def check_azure_remote_file_matching_local(
+    blob_properties: BlobProperties, local_file_path: Path
 ):
-    base_blob_name = remote_file_group["azure_base_blob_name"]
+    if not local_file_path.exists() or local_file_path.stat().st_size == 0:
+        return False
+
+    # Ask Azure for the md5 hash of the remote file.
+    content_settings = blob_properties.get("content_settings")
+    if not content_settings:
+        return False
+    remote_md5 = content_settings.get("content_md5")
+
+    # Compute the md5 hash of the local file.
+    with open(local_file_path) as file, mmap.mmap(
+        file.fileno(), 0, access=mmap.ACCESS_READ
+    ) as file:
+        local_md5 = hashlib.md5(file).digest()
+        return local_md5 == remote_md5
+
+
+def download_azure_remote_file(test_dir: Path, remote_file: str):
+    remote_file_name = remote_file.rsplit("/", 1)[-1]
+    relative_dir = test_dir.relative_to(THIS_DIR)
+
+    # Extract path components from Azure URL to use with the Azure Storage Blobs
+    # client library for Python (https://pypi.org/project/azure-storage-blob/).
+    #
+    # For example:
+    #   https://sharkpublic.blob.core.windows.net/sharkpublic/path/to/blob.txt
+    #                                            ^           ^
+    #   account_url:    https://sharkpublic.blob.core.windows.net
+    #   container_name: sharkpublic
+    #   blob_name:      path/to/blob.txt
+    #
+    # Note: we could also use the generic handler (e.g. wget, 'requests'), but
+    # the client library offers other APIs.
 
-    for remote_file in remote_file_group["files"]:
-        print(f"  Downloading {remote_file} to {test_dir.relative_to(THIS_DIR)}")
-        blob_name = base_blob_name + remote_file
-        dest = test_dir / remote_file
+    result = re.search(r"(https.+\.net)/([^/]+)/(.+)", remote_file)
+    account_url = result.groups()[0]
+    container_name = result.groups()[1]
+    blob_name = result.groups()[2]
 
-        with open(dest, mode="wb") as local_blob:
-            download_stream = container_client.download_blob(
-                blob_name, max_concurrency=4
-            )
+    with BlobClient(
+        account_url,
+        container_name,
+        blob_name,
+        max_chunk_get_size=1024 * 1024 * 32,  # 32 MiB
+        max_single_get_size=1024 * 1024 * 32,  # 32 MiB
+    ) as blob_client:
+        blob_properties = blob_client.get_blob_properties()
+
+        local_file_path = test_dir / remote_file_name
+        if check_azure_remote_file_matching_local(blob_properties, local_file_path):
+            print(f"  Skipping '{remote_file_name}' download (local MD5 hash matches)")
+            return
+
+        size_str = human_readable_size(blob_properties.size)
+        print(f"  Downloading '{remote_file_name}' ({size_str}) to '{relative_dir}'")
```

Comment:
Going to iterate on the logging here. This output is buffered when we want it to be streamed / flushed more regularly


https://github.com/nod-ai/SHARK-TestSuite/actions/runs/8708216214/job/23885060522?pr=164#step:6:1
```
Tue, 16 Apr 2024 15:19:19 GMT  python3 iree_tests/download_remote_files.py --root-dir pytorch/models
...
Tue, 16 Apr 2024 15:38:09 GMT Processing pytorch/models/opt-125M/test_cases.json
```